File: solutions/python/triangle/1/triangle.py
def triangle_inequality(sides):
    return sides[0]+sides[1] >= sides[2] and sides[0]+sides[2] >= sides[1] and sides[1]+sides[2] >= sides[0]

def all_zero(sides):
    return sides[0] == 0 and sides[1] == 0 and sides[2] == 0

def equilateral(sides):
    if not triangle_inequality(sides) or all_zero(sides):
        return False
    return sides[0] == sides[1] and sides[1] == sides[2]


def isosceles(sides):
    if not triangle_inequality(sides) or all_zero(sides):
        return False
    return sides[0] == sides[1] or sides[0] == sides[2] or sides [1] == sides[2]


def scalene(sides):
    if not triangle_inequality(sides) or all_zero(sides):
        return False
    return sides[0] != sides[1] and sides[0] != sides[2] and sides [1] != sides[2]

File: solutions/cpp/triangle/1/triangle.cpp
#include "triangle.h"

#include <stdexcept>

namespace triangle
{
    flavor kind(double a, double b, double c)
    {
        if (a <= 0 || b <= 0 || c <= 0)
        {
            throw std::domain_error("all sides have to be of length > 0");
        }

        if ((a + b) < c || (a + c) < b || (b + c) < a)
        {
            throw std::domain_error("the sum of the lengths of any two sides must be greater than or equal to the length of the third side");
        }

        if (a == b && a == c)
        {
            return flavor::equilateral;
        }

        else if (a == b || a == c || b == c)
        {
            return flavor::isosceles;
        }

        return flavor::scalene;
    }
} // namespace triangle
